### DISCUSSION TITLE
How to handle union types in query parameters?

### DISCUSSION THREAD

**Discussion by gloomybrain** [TypedProject/ts-express-decorators#1732] [Q&A] [answered]
Hey everyone! It seems I've got stuck with Ts.ED query parameter parsing. In particular I have a REST controller that accepts two types of the same query parameter:
 - `GET foo.bar/entity?createdAt[lte]=2021-12-10T10:10:17.717Z`
 - `GET foo.bar/entity?createdAt=2021-12-10T10:10:17.717Z`

The general idea to handle this is to have the following classes defined:
```typescript
class EntityDto {
  createdAt?: Date | EntityFilter<Date>;
}

class EntityFilter<T> {
  lte?: T;
}
```

As you can see I have to check whether the `createdAt` query parameter is a `String` and then make a `Date` instance out of it OR `createdAt` is an `Object` and I should create an instance of my `EntityFilter` class with the `lte` field set to an appropriate `Date` instance.

So my questions is as follows: what decorators should I use to define such a distinction? And how should my `EntityDto` class look like given that its createdAt field has to be either a Date or an EntityFilter instance?
I'd highly appreciate any help, and I suppose there might be some solutions other than writing these filters by hand.

**Comment by Romakita** [comment, marked as answer]
You can try to use the OnDeserialize decorator. It's not possible to automatically deserialize union type.

```typescript
@OnDeserialize(v => isObject(v) ? {lte: new Date(v.lte)} : new Date(v)) 
```

See you

**Reply by gloomybrain** [reply]
@Romakita thanks for the quick response!

Will try this

**Reply by gloomybrain** [reply]
Worked like charm, thank you again!

For those who are interested: it is still possible to make Ts.ED validate by providing schema-generating decorators, like this:
```typescript
@AdditionalProperties(false)
class EntityDto {
  @AnyOf(date().format('date-time'), from(EntityDateFilter) // this builds the validation json-schema
  @OnDeserialize(parseDateFilter) // and this performs the actual parsing of a raw query parameters object
  @Optional()
  public createdAt?: Date | EntityDateFilter;
}

@AdditionalProperties(false)
class EntityDateFilter {
  @Format('date-time')
  @Optional()
  gt?: Date; // means '>'
  lt?: Date; // means '<'
}

type RawDateFilter = {
  gt?: string;
  lt?: string;
};

const parseDateFilter = (value: string | RawDateFilter): Date | EntityDateFilter => {
  if (typeof value === 'string') {
    return new Date(value);
  }

  const filter = new EntityDateFilter();
  if (value.gt) {
    filter.gt = new Date(value.gt);
  }

  if (value.lt) {
    filter.lt = new Date(value.lt);
  }

  return filter;
}
```
